src/test/java/io/r2dbc/postgresql/client/BindingTest.java

```diff
@@ -52,7 +52,7 @@ void getParameterFormats() {
         binding.add(0, new Parameter(BINARY, 100, TEST.buffer(4).writeInt(200)));
         binding.add(2, new Parameter(BINARY, 100, TEST.buffer(4).writeInt(300)));
 
-        assertThat(binding.getParameterFormats()).containsExactly(BINARY, TEXT, BINARY);
```

Comment:
No, this shouldn't need to change and if these fail, it's a bug in the updated implementations.  What these are ensuring is that if you only bind parameters 0 and 2, then parameter 1 (which must exist) is contributed as an unspecified binding of some kind.  In your new implementation, you only iterate over the number of bindings that exist, potentially moving bindings from their specified index to an earlier one to fill in holes.

Reply:
Damn, I missed the fact you were skipping one.

I would think that this would give surprising results if left this way.

I would expect this to create an error instead of silently adding a TEXT format in the middle.

Reply:
Interesting.  That note in 51.2.3 seems to indicate that this is _possible_, but you think it's better that clients don't support the behavior.  I'm can definitely get on board with that if you'd like to update the implementation to behave that way.

Reply:
are you referring to 52.1.3 https://www.postgresql.org/docs/10/static/protocol-overview.html#PROTOCOL-FORMAT-CODES

If so I'm confused...

Reply:
OK, reading this again...
I understand the note in 51.2.3 to be saying. If you want the backend to infer the type of the parameter then set it to UNSPECIFIED (0). However by simply not binding anything as per the current test there is no Parameter.value. So yes, I would think this would be surprising to the user. Now if you want to  be able to have the server infer the type of the *value* then we should set the type to 0

On another note 51.2.3 refers to the 9.6 docs. I tend to use the latest which in a few days will be 11. I'm not sure how to deal with this, other than full href's

Reply:
Sounds good.  There are two options here then I think.  The first is to hold fire on this PR and start a second one to update existing behavior (pre-UNKNOWN) to throw an exception somewhere along the line if bindings are not contiguous.  The second option is to include that update as part of this change, which I'm OK with as the change without that (once it's been updated to encode-only) is quite small.  I'll leave it to you to decide how to go about it.

Reply:
OK, lets undo the changes to binding in this PR and commit what we agree on now, then fix the binding issue in another PR. History will be better served that way I believe.